### DISCUSSION TITLE
I'm having a problem with api guard

### DISCUSSION THREAD

**Discussion by MauricioBarbosa** [adonisjs/core#1944] [Help] [answered]
Always i try to use auth.use('api').authenticate(), it makes this query: 
`select * from api_tokens where d = 0 limit 1 - ER_BAD_FIELD_ERROR: Unknown column id in where clause`

But api_tokens doesn't have column id and it always tries to find 0. 

Someone can help? 

Here is my auth.ts file: 

`import { AuthConfig } from '@ioc:Adonis/Addons/Auth'

/*
|--------------------------------------------------------------------------
| Authentication Mapping
|--------------------------------------------------------------------------
|
| List of available authentication mapping. You must first define them
| inside the `contracts/auth.ts` file before mentioning them here.
|
*/
const authConfig: AuthConfig = {
  guard: 'web',
  list: {
    /*
    |--------------------------------------------------------------------------
    | Web Guard
    |--------------------------------------------------------------------------
    |
    | Web guard uses classic old school sessions for authenticating users.
    | If you are building a standard web application, it is recommended to
    | use web guard with session driver
    |
    */
    web: {
      driver: 'session',

      provider: {
        /*
        |--------------------------------------------------------------------------
        | Driver
        |--------------------------------------------------------------------------
        |
        | Name of the driver
        |
        */
        driver: 'lucid',

        /*
        |--------------------------------------------------------------------------
        | Identifier key
        |--------------------------------------------------------------------------
        |
        | The identifier key is the unique key on the model. In most cases specifying
        | the primary key is the right choice.
        |
        */
        identifierKey: 'usuario',

        /*
        |--------------------------------------------------------------------------
        | Uids
        |--------------------------------------------------------------------------
        |
        | Uids are used to search a user against one of the mentioned columns. During
        | login, the auth module will search the user mentioned value against one
        | of the mentioned columns to find their user record.
        |
        */
        uids: ['usuario'],

        /*
        |--------------------------------------------------------------------------
        | Model
        |--------------------------------------------------------------------------
        |
        | The model to use for fetching or finding users. The model is imported
        | lazily since the config files are read way earlier in the lifecycle
        | of booting the app and the models may not be in a usable state at
        | that time.
        |
        */
        model: () => import('App/Models/UsuarioAdministrador'),
      },
    },
    /*
    |--------------------------------------------------------------------------
    | OAT Guard
    |--------------------------------------------------------------------------
    |
    | OAT (Opaque access tokens) guard uses database backed tokens to authenticate
    | HTTP request. This guard DOES NOT rely on sessions or cookies and uses
    | Authorization header value for authentication.
    |
    | Use this guard to authenticate mobile apps or web clients that cannot rely
    | on cookies/sessions.
    |
    */
    api: {
      driver: 'oat',

      /*
      |--------------------------------------------------------------------------
      | Tokens provider
      |--------------------------------------------------------------------------
      |
      | Uses SQL database config for managing tokens. The foreignKey column is used
      | to make the relationship between the user and the token. You are free to
      | use any column name here.
      |
      */
      tokenProvider: {
        driver: 'database',
        table: 'api_tokens',
        foreignKey: 'user_id',
      },

      provider: {
        /*
        |--------------------------------------------------------------------------
        | Driver
        |--------------------------------------------------------------------------
        |
        | Name of the driver
        |
        */
        driver: 'lucid',

        /*
        |--------------------------------------------------------------------------
        | Identifier key
        |--------------------------------------------------------------------------
        |
        | The identifier key is the unique key on the model. In most cases specifying
        | the primary key is the right choice.
        |
        */
        identifierKey: 'cliente',

        /*
        |--------------------------------------------------------------------------
        | Uids
        |--------------------------------------------------------------------------
        |
        | Uids are used to search a user against one of the mentioned columns. During
        | login, the auth module will search the user mentioned value against one
        | of the mentioned columns to find their user record.
        |
        */
        uids: ['cliente'],

        /*
        |--------------------------------------------------------------------------
        | Model
        |--------------------------------------------------------------------------
        |
        | The model to use for fetching or finding users. The model is imported
        | lazily since the config files are read way earlier in the lifecycle
        | of booting the app and the models may not be in a usable state at
        | that time.
        |
        */
        model: () => import('App/Models/UsuarioCliente'),
      },
    },
  },
}

export default authConfig`

**Comment by thetutlage** [comment, marked as answer]
Share the code that you have used to generate the token and also how you are sending the token in request

**Reply by MauricioBarbosa** [reply]
Generate the Token

`public async registrar_cliente({response, request, auth}: HttpContextContract){
        var dados = request.all(); 

        const usuarioCliente = new UsuarioCliente();
        const cliente = new Cliente(); 

        cliente.nome = dados.cliente.nome; 
        cliente.cpf = dados.cliente.cpf;
        cliente.datanascimento = dados.cliente.datanascimento;
        cliente.sexo = dados.cliente.sexo;
        cliente.telefone = dados.cliente.telefone; 
        cliente.celular = dados.cliente.celular; 

        usuarioCliente.email = dados.usuario.email;
        usuarioCliente.senha = Md5.init(dados.usuario.senha);

        /*usuarioCliente.$attributes = dados.usuario; 
        cliente.$attributes = dados.cliente;  */
        console.log(usuarioCliente);
        console.log(cliente);

        try {
            await usuarioCliente.related('cliente').associate(cliente);
            const token = await auth.use('api').login(usuarioCliente, {
                expiresIn: '30 days'
            })
            response.status(200).json(token);
        } catch (error) {
            console.log(error);
            if(error.errno == 1062){
                response.status(500).json({erro: "E_DUP_USER", msg: "Já existe um usuário com esse e-mail"}); 
            }else{
                response.status(500).json({erro: "E_INT_ERROR", msg: "Erro ao inserir o usuário"}); 
            }
        }
   }`

`public async login({response, request, auth}: HttpContextContract){
        var dados = request.all(); 

        const usuarioCliente = new UsuarioCliente(); 
        usuarioCliente.email = dados.email; 
        usuarioCliente.senha = Md5.init(dados.senha); 

        try{
            const token = await auth.use('api').login(usuarioCliente);
            response.status(200).json(token);
        }catch(erro){
            console.log('erro');
        }
    }
`

**Reply by MauricioBarbosa** [reply]
I'm also sending it on headers : 

`

HOST | 127.0.0.1:3333
-- | --
USER-AGENT | insomnia/2020.4.2
CONTENT-TYPE | application/json
AUTHORIZATION | Bearer MA.OZqZGRF-GFVKg3gHDaiRDLUwNfxOUUNed7LpjKy8LrPmhERA9kgdxcl1cbkc
ACCEPT | */*


`

**Reply by MauricioBarbosa** [reply]
?

**Comment by MauricioBarbosa** [comment]
My problem was solved when I set id in api_tokens as AI

**Reply by paulorobsant** [reply]
Hello. Can you explain more your solution?

If you are brazilian you can just write in portuguese too, please?

**Reply by pimentelleo** [reply]
Eu acredito que ele mudou a coluna dos ids do api token para AI, não sei ao certo como isso resolveria mas deve ser pela lógica do código dele
